packages/react-theme-provider/src/components/ThemeProvider/useThemeStyleTag.ts

```diff
@@ -0,0 +1,47 @@
+import * as React from 'react';
+import { useIsomorphicLayoutEffect, useId } from '@fluentui/react-utilities';
+import { themeToCSSVariables } from '@fluentui/react-theme';
+import { ThemeProviderState } from './ThemeProvider.types';
+
+/**
+ * Writes a theme as css variables in a style tag on the provided document as a rule applied to a CSS class
+ *
+ * @param theme Theme values that are written as css variables in style rule
+ * @returns CSS class to apply the rule
+ */
+export const useThemeStyleTag = (theme: ThemeProviderState['theme'], document: ThemeProviderState['document']) => {
+  const styleTagId = useId('theme-provider');
+
+  const cssRule = React.useMemo(() => {
+    const cssVars = themeToCSSVariables(theme);
+    const cssVarsAsString = Object.keys(cssVars).reduce((cssVarRule, cssVar) => {
+      cssVarRule += `${cssVar}: ${cssVars[cssVar]}; `;
+      return cssVarRule;
+    }, '');
+
+    // result: .theme-provider { --css-var: '#fff' }
+    return `.${styleTagId} { ${cssVarsAsString} }`;
+  }, [theme, styleTagId]);
+
+  useIsomorphicLayoutEffect(() => {
+    if (document === null || document === undefined) {
+      return;
+    }
+
+    const styleTag = document.createElement('style');
+    styleTag.setAttribute('id', styleTagId);
+    document.head.appendChild(styleTag);
+    (styleTag.sheet as CSSStyleSheet).insertRule(cssRule, 0);
+
+    // Removes the style tag from the document
+    return () => {
+      const styleTagCleanup = document?.getElementById(styleTagId);
+      if (styleTagCleanup) {
+        // IE11 safe node removal, otherwise use node.remove()
+        styleTagCleanup.parentElement?.removeChild(styleTagCleanup);
+      }
+    };
+  }, [styleTagId, cssRule]);
```

Comment:
I think that we should also modify the logic there: currently it will remove & create a style node on `cssRule` changes, IMO we should simply replace a CSS rule.

Reply:
I think that should be done now but I am no expert with the CSSStylesheet API 😅. Please review that logic and resolve this comment if it looks OK